### ISSUE TITLE
Tokenizer Comparison: CNN (13M) achieves 30+ PSNR vs Transformer (40M) at 27 PSNR

### ISSUE THREAD

**Issue by daniyal-rahman** [daniyal-rahman/ahriuwu#6]
## Summary

Compared three tokenizer architectures for frame reconstruction quality:

| Model | Params | PSNR (↑) | LPIPS (↓) | MSE (↓) |
|-------|--------|----------|-----------|---------|
| **CNN** | **13M** | **32.64 ± 0.87 dB** | **0.0216** | **0.00056** |
| Transformer (dynamics checkpoint) | 40M | 26.97 ± 1.22 dB | 0.1149 | 0.00209 |
| Transformer (current training) | 40M | 27.45 ± 1.29 dB | 0.1032 | 0.00188 |

## Key Findings

1. **CNN achieves 30+ PSNR** with only 1/3 the parameters of the transformer
2. **CNN LPIPS is 5x better** (0.02 vs 0.10) - perceptually much sharper
3. **Transformer PSNR has plateaued** at ~27 dB after 4 epochs, not improving significantly
4. Visual inspection confirms CNN reconstructions are notably sharper than transformer
5. **Neither tokenizer uses action labels** - both are pure image→latent→image autoencoders, so comparison is fair

## Visual Comparisons

Each image shows: `Original | CNN | Trans-Dyn | Trans-Cur` with PSNR and LPIPS metrics.

### Sample 1 - Standard gameplay
![Comparison 0](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_000.png)

### Sample 2 - Team fight
![Comparison 5](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_005.png)

### Sample 3 - Jungle area
![Comparison 10](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_010.png)

### Sample 4 - Lane phase
![Comparison 2](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_002.png)

### Sample 5 - UI heavy
![Comparison 7](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_007.png)

### Sample 6 - Dark scene
![Comparison 15](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_015.png)

<details>
<summary>All 20 comparison images</summary>

![0](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_000.png)
![1](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_001.png)
![2](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_002.png)
![3](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_003.png)
![4](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_004.png)
![5](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_005.png)
![6](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_006.png)
![7](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_007.png)
![8](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_008.png)
![9](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_009.png)
![10](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_010.png)
![11](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_011.png)
![12](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_012.png)
![13](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_013.png)
![14](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_014.png)
![15](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_015.png)
![16](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_016.png)
![17](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_017.png)
![18](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_018.png)
![19](https://raw.githubusercontent.com/daniyal-rahman/ahriuwu/main/samples/tokenizer_comparison/comparison_019.png)

</details>

### Why is CNN better despite fewer parameters?

Potential reasons:
1. **Architecture mismatch**: Transformers excel at global context but may over-smooth local details
2. **MAE training**: Transformer trained with masking (0-90%) which may hurt peak reconstruction quality
3. **Loss function**: Both use MSE + LPIPS, but transformer has higher baseline LPIPS
4. **Latent space**: CNN uses dense conv features vs transformer's patch-based tokenization

### Qualitative observations
- CNN preserves fine details (UI elements, character models, particle effects)
- Transformer reconstructions appear blurry/smoothed
- Transformer struggles more with high-frequency details (text, health bars)

### Implications for World Model

- For **reconstruction quality**: CNN is clearly superior
- For **dynamics modeling**: Transformer's patch-based latents may still be better for temporal prediction
- **Trade-off**: May need separate tokenizers for reconstruction vs dynamics

## Recommendations

1. **For 30+ PSNR target**: Use CNN tokenizer (already achieves this)
2. **Investigate transformer issues**:
   - Try training transformer WITHOUT masking (pure autoencoder)
   - Try reducing LPIPS weight (optimize more for PSNR)
   - Try larger model (medium/large) to see if capacity is the limit
3. **Consider hybrid approach**: CNN encoder → Transformer dynamics → CNN decoder

## Test Configuration

- 20 random frames sampled from `data/processed/frames` (2.3M total)
- Seed: 42
- Device: CUDA with bfloat16 autocast
- Checkpoints:
  - CNN: `checkpoints/tokenizer_epoch_010.pt`
  - Trans-Dyn: `checkpoints/run_20260115_195915/transformer_tokenizer_best.pt`
  - Trans-Cur: `checkpoints/run_20260117_165401/transformer_tokenizer_latest.pt`

## Reproduction

```bash
python scripts/compare_tokenizers.py --num-samples 20
```

Results: `samples/tokenizer_comparison/results.json`
